### PR TITLE
port fsih to fsi as a hash directive

### PR DESCRIPTION
**Pull request by dawedawe**
## Description

This is a port of [fsih](https://github.com/dawedawe/fsih) to fsi itself as a hash directive.

image

I think using a hash directive for this makes the most sense but as we use the normal parser for them it also means we have to wrap the expression in parentheses. I hope that's an acceptable tradeoff.
If you want to give this a try, you have to copy the `FSharp.Core.xml` manually to the fsi artifacts folder. The build process doesn't do it. Later, in the SDK distribution, the xml file will be there as far as I can see.

I want to mention that this PR was made possible by the supporters of the [Amplifying FSharp Open Collective](https://opencollective.com/amplifying-fsharp).

## Checklist

- [x] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [X] Release notes entry updated:
    > Please make sure to add an entry with short succinct description of the change as well as link to this pull request to the respective release notes file, if applicable.
    >
    > Release notes files:
    > - If anything under `src/Compiler` has been changed, please make sure to make an entry in `docs/release-notes/.FSharp.Compiler.Service/<version>.md`, where `<version>` is usually "highest" one, e.g. `42.8.200`
    > - If language feature was added (i.e. `LanguageFeatures.fsi` was changed), please add it to `docs/releae-notes/.Language/preview.md`
    > - If a change to `FSharp.Core` was made, please make sure to edit `docs/release-notes/.FSharp.Core/<version>.md` where version is "highest" one, e.g. `8.0.200`.

    > Information about the release notes entries format can be found in the [documentation](https://fsharp.github.io/fsharp-compiler-docs/release-notes/About.html).
    > Example:
    > * More inlines for Result module. ([PR #16106](https://github.com/dotnet/fsharp/pull/16106))
    > * Correctly handle assembly imports with public key token of 0 length. ([Issue #16359](https://github.com/dotnet/fsharp/issues/16359), [PR #16363](https://github.com/dotnet/fsharp/pull/16363))
    > *`while!` ([Language suggestion #1038](https://github.com/fsharp/fslang-suggestions/issues/1038), [PR #14238](https://github.com/dotnet/fsharp/pull/14238))

    > **If you believe that release notes are not necessary for this PR, please add `NO_RELEASE_NOTES` label to the pull request.**

### CODE REVIEW

**Comment by vzarytovskii** [comment]
Nice, I'm fine with having it as default, it will need tests though (simple baseline which verifies couple of outputs).

And an approval from @KevinRansom

**Comment by baronfel** [comment]
From a usability perspective, is there any auto complete support for hash directive contents? If so, users could dot-into the expression they want docs for.

**Comment by psfinaki** [comment]
Hey Dawe, I think this would be a great addition to FSI, worth spreading a word about it. 

A note on errh clean code, would you want to move the new module to a separate file or something like this? `fsi.fs` is already very fat and seems like a rather easy target for extraction.

File: src/Compiler/Interactive/fsi.fs
Comment (vzarytovskii) [inline src/Compiler/Interactive/fsi.fs:1417]: Probably should be more verbose, something like `FsHelp` or something along the lines.

**Comment by smoothdeveloper** [comment]
Considerations:

* would it be better done as a FSI extension through `#r "fsih: List.map"`?
* for terminal output, it would be possible to pretty print a bit the output, using brighter coloring for headings such as "Description:", "Parameters:", etc. as well as bulleted items, it will improve the legibility of the output; or maybe in markdown format (the IDE can then figure out how to pretty print when handling the evaluation)
* the code samples would also have syntax colouring (probably not semantic coloring that requires type checking, but whatever we could have), but this is probably more challenging item

Disregarding this, it already sound like nice feature that will speak to users of python REPL and other similar environments 🙂.

**Comment by vzarytovskii** [comment]
> Considerations:
> 
> * would it be better done as a FSI extension through `#r "fsih: List.map"`?

It's not a dependency manager related functionality, so probably not(?).

> * for terminal output, it would be possible to pretty print a bit the output, using brighter coloring for headings such as "Description:", "Parameters:", etc. as well as bulleted items, it will improve the legibility of the output; or maybe in markdown format (the IDE can then figure out how to pretty print when handling the evaluation)

In this case, we will need additional testing for it for different `$TERM`s, and that it's not breaking the output of interactive and when used as API.

> * the code samples would also have syntax colouring (probably not semantic coloring that requires type checking, but whatever we could have), but this is probably more challenging item
> 

Also would require much more testing with both non-terminal environments as well as all popular emulators and shells.

**Comment by brianrourkeboll** [comment]
> From a usability perspective, is there any auto complete support for hash directive contents?

There is autocomplete in VS for filepaths inside the `"…"` after `#I "…"`, `#r "…"`, etc., in an `.fsx` file.

image

But I think I'm a bit thrown off by the `#h "expr";;` syntax myself.

What is the primary motivation for using a hash directive?

1. For symmetry with `#help`?
2. So that this feature itself shows up in `#help` (which currently only shows help for itself and other FSI directives)?
3. To ensure that the user can't accidentally shadow the ability to access this functionality by defining their own value `h`?

(1) doesn't feel super-compelling to me, since all the other FSI directives, including `#help`, operate on the level of (i.e., refer to or affect) the FSI session, not on expressions within the session.

If (2), then I think we could just add the help for `h` to the information that `#help` shows. We could also add help for the [existing `fsi` object](https://learn.microsoft.com/en-us/dotnet/fsharp/tools/fsharp-interactive/#using-the-fsi-object-in-f-code) while we were at it.

If (3), then what about just adding an `h` method to the existing `fsi` object, which is less likely to be accidentally shadowed than `h` on its own?

```fsx
fsi.h List.map;;
```

**Comment by dawedawe** [comment]
> Hey Dawe, I think this would be a great addition to FSI, worth spreading a word about it once it's in.
> 
> A note on errh clean code, would you want to move the new module to a separate file or something like this? `fsi.fs` is already very fat and seems like a rather easy target for extraction.

Thanks :)
Yeah, I thought about that but then decided to follow the existing all-in-one-file style of fsi for consistency.
But sure, if y'all are okay with a dedicated file, I'll definitely split it off.

**Comment by dawedawe** [comment]
> What is the primary motivation for using a hash directive?
> 
>     1. For symmetry with `#help`?
> 
>     2. So that this feature itself shows up in `#help` (which currently only shows help for itself and other FSI directives)?
> 
>     3. To ensure that the user can't accidentally shadow the ability to access this functionality by defining their own value `h`?

My motivation was an uniform distribution among all three points.
Avoiding shadowing is a must-have for me.

You're making a strong argument for a new `fsi.h` function to provide the feature. But users should be able to discover it in the output of `#help`. So yeah, as long as the team is cool with it, that would also be my preferred way forward.

**Comment by smoothdeveloper** [comment]
> #r is a dependency manager functionality/extension. And if done via DM, it will be quite verbose and confusing.

I concede it might sound unatural, but #r stands for "reference" (looking for "reference documentation"), and #r extensions can/do print to the output. It would need 0 parser changes.

@brianrourkeboll suggestion of `fsi.h` sounds much better in any case and very close to current `fsih` package.

I also prefer it to hash directive and quoting the expression.

Are there special events that FSI session will trigger?

Can such events be done from a type extension over `FSharp.Compiler.Interactive.InteractiveSession`?

This could help extending IDEs in terms of how the output is parsed / displayed (with the idea of markdown or pretty printing strategies that could be implemented on top, rather than raw parsing of stdout).

**Comment by dawedawe** [comment]
Okay, I moved to `fsi.h` as an user interface and added a printer for the output processing. Meaning, the record option containing the various documentation pieces is available as `it` afterwards.
image

The downside is, that the `FSharp.Compiler.Interactive.Settings` project (which hosts the `fsi` aka `InteractiveSession`) can't access the shim filesystem.

File: src/FSharp.Compiler.Interactive.Settings/fsihelp.fs
Comment (brianrourkeboll) [inline src/FSharp.Compiler.Interactive.Settings/fsihelp.fs:5]: These assembly-level attributes are also in `fsiaux.fs`:

https://github.com/dotnet/fsharp/blob/a7a2d4ae2f88b62b7d2de778c605c299de158f30/src/FSharp.Compiler.Interactive.Settings/fsiaux.fs#L12-L14

But I think they only need to be in one place, so you could remove them from there, since this is now the first file in the project.

You could also just put them in an `AssemblyInfo.fs` file, or [add them to the fsproj](https://learn.microsoft.com/dotnet/standard/assembly/set-attributes-project-file#set-arbitrary-attributes):

```fsproj
<ItemGroup>
  <AssemblyAttribute Include="System.Runtime.InteropServices.ComVisible">
    <_Parameter1>false</_Parameter1>
  </AssemblyAttribute>
  <AssemblyAttribute Include="System.CLSCompliant">
    <_Parameter1>true</_Parameter1>
  </AssemblyAttribute>
</ItemGroup>
```

File: src/FSharp.Compiler.Interactive.Settings/fsihelp.fs
Comment (brianrourkeboll) [inline src/FSharp.Compiler.Interactive.Settings/fsihelp.fs:37]: These could also be something like

```suggestion
            sb.AppendLine()
                .AppendLine("Description:")
                .AppendLine(this.Summary)
                |> ignore
```

(It might be easier to keep track of what pieces have their own line that way.)

File: src/FSharp.Compiler.Interactive.Settings/fsiaux.fs
Comment (dawedawe) [inline src/FSharp.Compiler.Interactive.Settings/fsiaux.fs:186]: Does someone know why the `None` is not triggered? Something with the hosting  and reflection mechanism of the session?

File: src/FSharp.Compiler.Interactive.Settings/fsihelp.fs
Comment (brianrourkeboll) [inline src/FSharp.Compiler.Interactive.Settings/fsihelp.fs:94]: Should this just be `tryGetXmlDocument`, with `try Some … with _ -> None`? Then you don't need the `File.Exists` in `getInfos`. In theory the file could disappear or become inaccessible between that call and this one anyway...

File: src/FSharp.Compiler.Interactive.Settings/fsihelp.fs
Comment (brianrourkeboll) [inline src/FSharp.Compiler.Interactive.Settings/fsihelp.fs:255]:
```suggestion
        | NewTuple _ ->
            let ty = typeof<_ * _>
            getInfos ty (Some ty.Name) ty.Name
        | NewStructTuple _ ->
            let ty = typeof<struct (_ * _)>
            getInfos ty (Some ty.Name) ty.Name
```

File: src/FSharp.Compiler.Interactive.Settings/fsiaux.fs
Comment (brianrourkeboll) [reply src/FSharp.Compiler.Interactive.Settings/fsiaux.fs:186]: It seems to do that for any `'T option`:

```fsx
type T = T of int
fsi.AddPrinter (function Some (T t) -> $"Got a T: {t}." | None -> "Nope lol.")
let some = Some (T 1)
let none = Option<T>.None
```

```fsi
type T = | T of int
val some: T option = Got a T: 1.
val none: T option = None
```

It works for `'T voption`, though:

```fsx
type T = T of int
fsi.AddPrinter (function ValueSome (T t) -> $"Got a T: {t}." | ValueNone -> "Nope lol.")
let some = ValueSome (T 1)
let none = ValueOption<T>.ValueNone
```

```fsi
type T = | T of int
val some: T voption = Got a T: 1.
val none: T voption = Nope lol.
```

Comment (dawedawe) [reply src/FSharp.Compiler.Interactive.Settings/fsiaux.fs]: Awesome, thanks a ton for figuring this out. We should get this documented.

File: src/FSharp.Compiler.Interactive.Settings/fsihelp.fs
Comment (dawedawe) [reply src/FSharp.Compiler.Interactive.Settings/fsihelp.fs:94]: 👍 
Will try tomorrow. It's getting late here...

File: src/FSharp.Compiler.Interactive.Settings/fsihelp.fs
Comment (dawedawe) [reply src/FSharp.Compiler.Interactive.Settings/fsihelp.fs:5]: I kept the single attributes in code as I'd like to leave it to the team to decide if we should put them in an `AssemblyInfo.fs or the fsproj file.

**Comment by dawedawe** [comment]
I think this is in an okay state now but I don't know how to add tests for it.
Is there some testing facility that can actually access the `fsi` object?
I tried `FSharpScript` from `ScriptHelpers.fs` but it seems like it can only do hash directives and normal code, no member calls of `fsi`.

**Comment by baronfel** [comment]
Maybe just test the implementation directly with various `Expr`s? I'm not sure there's value in testing that you can call `fsi.h` directly, it's more important that the behavior of the expression-traversal be covered by tests.

**Review by edgarfgp — APPROVED** [review]

File: src/FSharp.Compiler.Interactive.Settings/fsihelp.fsi
Comment (psfinaki) [inline src/FSharp.Compiler.Interactive.Settings/fsihelp.fsi:21]: tryGetHelp maybe?

**Review by psfinaki — COMMENTED** [review]
Good stuff. I personally liked the "#h" option more but don't have a hard preference here. 

I think it would be good to consolidate "help" vs "documentation" terminology here, it's essentially the same in this context AFAIU. Otherwise the implementation looks solid to me.

**Comment by KevinRansom** [comment]
Okay, we like the idea of this PR, but perhaps not the experience:
A func on the session fsi object is not really ideal

Fsi already has a #help command which does something, and is probably the likeliest guess a naive user is going to make.
````
Microsoft (R) F# Interactive version 12.8.400.0 for F# 8.0
Copyright (c) Microsoft Corporation. All Rights Reserved.
For help type #help;;

> #help
- ;;

  F# Interactive directives:

    #r "file.dll";;                               // Reference (dynamically load) the given DLL
    #i "package source uri";;                     // Include package source uri when searching for packages
    #I "path";;                                   // Add the given search path for referenced DLLs
    #load "file.fs" ...;;                         // Load the given file(s) as if compiled and referenced
    #time ["on"|"off"];;                          // Toggle timing on/off
    #help;;                                       // Display help
    #r "nuget:FSharp.Data, 3.1.2";;               // Load Nuget Package 'FSharp.Data' version '3.1.2'
    #r "nuget:FSharp.Data";;                      // Load Nuget Package 'FSharp.Data' with the highest version
    #clear;;                                      // Clear screen
    #quit;;                                       // Exit

  F# Interactive command line options:

      See 'dotnet fsi --help' for options
>````

extending this to also support: ````#help "System.Console.In";;```` should be trivial.

As observed, #help "System.Console.In";;  with quotes is a bit irritating to this end we have proposed this PR: https://github.com/dotnet/fsharp/pull/17206 which will remove the necessity of quotes, allowing long idents to be retrieved.

So I believe we will be in favour of merging functionality similar to this with the above experience changes.

Thanks

Kevin

**Comment by dawedawe** [comment]
Thanks for the reviews @psfinaki and @KevinRansom,
once #17206 is merged, I'll rework this PR to use `#help` as Kevin suggested.

**Comment by KevinRansom** [comment]
@dawedawe, no need to wait mate.  You can make it work with quotes, while we work the language process.  When the other PR is merged it should either just work, or be a trivial fix.

**Comment by dawedawe** [comment]
@KevinRansom Cool, so this is the state now:
Screenshot from 2024-06-06 12-02-30

**Review by T-Gro — APPROVED** [review]

File: tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs
Comment (KevinRansom) [inline tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs:722]: I dislike bike shedding and words are super hard.  But (gonna do it anyway :-) ) this isn't really an expression.  I think either symbol, member or identifier may be a better word.  I think my preference is identifier, but am open to better suggestions.

**Review by KevinRansom — CHANGES_REQUESTED** [review]
Nice, I think this will be useful.  I have one issue about the help text, it's not really an expression, it is probably better described as an identifier.

File: tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs
Comment (dawedawe) [reply tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs]: Good point, I changed it to "identifier".

File: src/Compiler/Interactive/fsihelp.fs
Comment (psfinaki) [inline src/Compiler/Interactive/fsihelp.fs:264]: A question about the unhappy path here - what is the case for it? You can answer with a test :)

**Review by KevinRansom — APPROVED** [review]
Nice work.

Thanks

**Comment by KevinRansom** [comment]
Thanks for this.

File: tests/fsharp/core/printing/output.stdout.bsl
Comment (psfinaki) [reply tests/fsharp/core/printing/output.stdout.bsl:1]: Thanks!

**Pull request merged**
